{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File Name and Extension #1: tor.txt\n",
      "File Not Found. Enter Valid File Name! \n",
      "\n",
      "File Name and Extension #1: tor2.txt\n",
      "File Name and Extension #2: wuhan.txt\n",
      "File Name and Extension #3: urbani.txt\n",
      "Number of Nucleotides to Align: 100\n"
     ]
    }
   ],
   "source": [
    "#opens the each file and reads it and asks how many nucleotides to align\n",
    "x = False\n",
    "while x == False: \n",
    "    try: \n",
    "        #tries to open each of the files\n",
    "        file1 = open(input (\"File Name and Extension #1: \"), \"r\")\n",
    "        file2 = open(input (\"File Name and Extension #2: \"), \"r\")\n",
    "        file3 = open(input (\"File Name and Extension #3: \"), \"r\")\n",
    "        x = True\n",
    "    except: \n",
    "        print(\"File Not Found. Enter Valid File Name! \\n\")\n",
    "\n",
    "y = False\n",
    "while y == False: \n",
    "    try: \n",
    "        #gets the first x nucleotides to align\n",
    "        number = int(input(\"Number of Nucleotides to Align: \"))\n",
    "        y = True\n",
    "    except:\n",
    "        print(\"Enter a number!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#makes the sequences into one line for easy readability\n",
    "file1.readline()\n",
    "seq1 = \"\\n\".join(file1.readlines())[:number]\n",
    "seq1 = seq1.replace(\"\\n\",\"\")\n",
    "\n",
    "file2.readline()\n",
    "seq2 = \"\\n\".join(file2.readlines())[:number]\n",
    "seq2 = seq2.replace(\"\\n\",\"\")\n",
    "\n",
    "file3.readline()\n",
    "seq3 = \"\\n\".join(file3.readlines())[:number]\n",
    "seq3 = seq3.replace(\"\\n\",\"\")\n",
    "\n",
    "#closes the files\n",
    "file1.close()\n",
    "file2.close()\n",
    "file3.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations \n",
    "#msa\n",
    "sequences = [seq1, seq2, seq3]\n",
    "\n",
    "#calcualtes the combinations of 2\n",
    "seqs = combinations(sequences, 2) \n",
    "\n",
    "#variables to stores the outputs from the pairwise allignment\n",
    "seq_info = []\n",
    "identities = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "#Dynamic Sequencing\n",
    "def dynamic_seq(seq1, seq2):\n",
    "    #creates the matrixes\n",
    "    matrix = np.zeros((len(seq2)+1, len(seq1)+1))\n",
    "    mutation = []\n",
    "    for row in range(0, len(seq2)+1):\n",
    "        values = []\n",
    "        for col in range (0, len(seq1)+1):\n",
    "            values.append([])\n",
    "        mutation.append(values)\n",
    "\n",
    "\n",
    "\n",
    "    #scores for Needleman gap, mismatch, match\n",
    "    gap = -2\n",
    "    match = 1\n",
    "    mismatch = -1\n",
    "\n",
    "\n",
    "    #first fill in the first row and column\n",
    "    for row in range(1, len(seq2)+1):\n",
    "        matrix[row][0] = matrix [row-1][0]+gap\n",
    "        mutation [row][0].append(\"del\")\n",
    "    for col in range(1, len(seq1)+1):\n",
    "        matrix[0][col] = matrix [0][col-1]+gap\n",
    "        mutation [0][col].append(\"ins\")\n",
    "\n",
    "    #diagonal in the rest of the entries\n",
    "    for row in range (1, len(seq2)+1):\n",
    "        for col in range (1, len(seq1)+1):\n",
    "            value1 = 0\n",
    "            symbol = \"\"\n",
    "            #calculates values for match/mismatch\n",
    "            if seq1[col-1] == seq2[row-1]:\n",
    "                value1 = matrix[row-1][col-1] + match\n",
    "                symbol = \"match\"\n",
    "            else:\n",
    "                value1 = matrix[row-1][col-1] + mismatch\n",
    "                symbol = \"mismatch\"\n",
    "            #calculates values for gaps\n",
    "            value2 = matrix [row-1][col]+gap\n",
    "            value3 = matrix [row][col-1]+gap\n",
    "            matrix[row][col] = max(value1, value2, value3)\n",
    "\n",
    "            #assigns the symbol in the direction for where the value came from\n",
    "            if matrix[row][col] == value1:\n",
    "                mutation[row][col].append(symbol)\n",
    "            elif matrix[row][col] == value2:\n",
    "                mutation[row][col].append(\"del\")\n",
    "            elif matrix[row][col] == value3:\n",
    "                mutation[row][col].append(\"ins\")\n",
    "    return mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculates the pairwise allignment \n",
    "def pairwise_allignment (seq1, seq2, mutation):\n",
    "    align1 = \"\"\n",
    "    align2 = \"\"\n",
    "    final = \"\"\n",
    "    i = len(seq2)\n",
    "    j = len(seq1)\n",
    "    \n",
    "    #detemines the mutation and aligns the two sequences\n",
    "    while i > 0 and j > 0:\n",
    "        symbols = mutation[i][j]\n",
    "        for symbol in symbols:\n",
    "            if symbol == \"mismatch\":\n",
    "                align1 += seq2 [i-1]\n",
    "                align2 += seq1 [j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "                final += \"*\"\n",
    "            elif symbol == \"match\":\n",
    "                align1 += seq2 [i-1]\n",
    "                align2 += seq1 [j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "                final += \" \"\n",
    "            elif symbol == \"del\":\n",
    "                align1 += seq2[i-1]\n",
    "                align2 += \"-\"\n",
    "                i -= 1\n",
    "                final += \"*\"\n",
    "            elif symbol == \"ins\":\n",
    "                align1 += \"-\"\n",
    "                align2 += seq1[j-1]\n",
    "                j -= 1\n",
    "                final += \"*\"\n",
    "   \n",
    "    #completes allignment in the ends\n",
    "    while j > 0:\n",
    "        align1 += \"-\"\n",
    "        align2 += seq2[j-1]\n",
    "        j -= 1\n",
    "        final += \"*\"\n",
    "    while i > 0:\n",
    "        align1 += seq1[i-1]\n",
    "        align2 += \"-\"\n",
    "        i -= 1\n",
    "        final += \"*\"\n",
    "\n",
    "    #reorients the alignments\n",
    "    align1 = align1 [::-1]\n",
    "    align2 = align2 [::-1]\n",
    "    final = final [::-1]\n",
    "    \n",
    "    #calcualtes the different percent_identities\n",
    "    percent_identity = round(final.count (\" \")/len(final)*100,2)\n",
    "    return {\"align1\": align1,\"align2\": align2, \"final\": final, \"percent_identity\":percent_identity}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#once you get the combination, does the allignments and assigns the percent identities to identities\n",
    "for seq in seqs: \n",
    "    mut = dynamic_seq(seq[0], seq[1])\n",
    "    info = pairwise_allignment (seq[0], seq[1], mut)\n",
    "    seq_info.append(info)\n",
    "    identities.append(info.get(\"percent_identity\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statistics import median\n",
    "\n",
    "#does the last two alignment\n",
    "last_alignment = seq_info[identities.index(median(identities))]\n",
    "\n",
    "#hard-coded ATT which is the first three nucleotides in Wuhan strain\n",
    "##reason for having to start with the 1st nucleotide in this code; \n",
    "##why we did not do 3 different msa's each of 10,000 base pairs\n",
    "#gets the alginment for the Tor2 strain and the Wuhan strain\n",
    "if seq_info[identities.index(median(identities))].get(\"align1\")[:3] == \"ATT\":\n",
    "    tor2 = seq_info[identities.index(median(identities))].get(\"align2\")\n",
    "    wuhan = seq_info[identities.index(median(identities))].get(\"align1\")\n",
    "else: \n",
    "    tor2 = seq_info[identities.index(median(identities))].get(\"align1\")\n",
    "    wuhan = seq_info[identities.index(median(identities))].get(\"align2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determines which of the sequence is the Urbani sequence after doing the alignment\n",
    "urbani_1 = seq_info[identities.index(max(identities))].get(\"align1\")\n",
    "urbani_2 = seq_info[identities.index(max(identities))].get(\"align2\")\n",
    "\n",
    "#does the two alignments with both the potential urbani strains\n",
    "first_alignment_one = dynamic_seq(tor2, urbani_1)\n",
    "first_align_info_one = pairwise_allignment(tor2, urbani_1, first_alignment_one)\n",
    "first_alignment_two = dynamic_seq(tor2, urbani_2)\n",
    "first_align_info_two = pairwise_allignment(tor2, urbani_2, first_alignment_two)\n",
    "\n",
    "#determines highest percent identity of the two alignments\n",
    "if first_align_info_one.get(\"percent_identity\") > first_align_info_two.get(\"percent_identity\"):\n",
    "    first_align_info = first_align_info_two\n",
    "else: \n",
    "    first_align_info = first_align_info_one \n",
    "\n",
    "#the urbani strain is determined\n",
    "if first_align_info.get(\"align1\") == tor2:\n",
    "    urbani = first_align_info.get(\"align2\")\n",
    "else: \n",
    "    urbani = first_align_info.get(\"align1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculates the final alignment\n",
    "i = 0\n",
    "final = \"\"\n",
    "final1 = first_align_info.get(\"final\")\n",
    "final2 = last_alignment.get(\"final\")\n",
    "while i < len(final1):\n",
    "    if final1[i] == \"*\" or final2[i] == \"*\":\n",
    "        final = final + \"*\"\n",
    "    else:\n",
    "        final = final + \" \"\n",
    "    i += 1\n",
    "\n",
    "#calculates percent identity\n",
    "percent_identity = round(final.count(\" \")/len(final)*100,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#creates an output file that aligns the three sequences together\n",
    "output = open(\"output.txt\", \"w\")\n",
    "i = 0\n",
    "\n",
    "#determines reminder for the last line\n",
    "reminder = len(final) % 50\n",
    "\n",
    "#prints the allignments in sets of 50\n",
    "while i < len(final)-reminder:\n",
    "    output.write(\"Urbani: \" + urbani[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Tor2:   \" + tor2[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Wuhan:  \" + wuhan[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Final:  \" + final[i:i+50]+\" \"+ str(i+50)+ \"\\n \\n\")\n",
    "    i = i + 50\n",
    "\n",
    "#prints the last lines of the allignment\n",
    "output.write(\"Urbani: \" + urbani[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Tor2:   \" + tor2[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Wuhan:  \" + wuhan[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Final:  \" + final[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n \\n\")\n",
    "\n",
    "#prints the percent identity\n",
    "output.write(\"Percent Identity: \" + str(percent_identity) + \"%\")\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#opens the each file and reads it and asks how many nucleotides to align\n",
    "x = False\n",
    "while x == False: \n",
    "    try: \n",
    "        #tries to open each of the files\n",
    "        file1 = open(input (\"File Name and Extension #1: \"), \"r\")\n",
    "        file2 = open(input (\"File Name and Extension #2: \"), \"r\")\n",
    "        file3 = open(input (\"File Name and Extension #3: \"), \"r\")\n",
    "        x = True\n",
    "    except: \n",
    "        print(\"File Not Found. Enter Valid File Name! \\n\")\n",
    "\n",
    "y = False\n",
    "while y == False: \n",
    "    try: \n",
    "        #gets the first x nucleotides to align\n",
    "        number = int(input(\"Number of Nucleotides to Align: \"))\n",
    "        y = True\n",
    "    except:\n",
    "        print(\"Enter a number!\")\n",
    "\n",
    "#makes the sequences into one line for easy readability\n",
    "file1.readline()\n",
    "seq1 = \"\\n\".join(file1.readlines())[:number]\n",
    "seq1 = seq1.replace(\"\\n\",\"\")\n",
    "\n",
    "file2.readline()\n",
    "seq2 = \"\\n\".join(file2.readlines())[:number]\n",
    "seq2 = seq2.replace(\"\\n\",\"\")\n",
    "\n",
    "file3.readline()\n",
    "seq3 = \"\\n\".join(file3.readlines())[:number]\n",
    "seq3 = seq3.replace(\"\\n\",\"\")\n",
    "\n",
    "#closes the files\n",
    "file1.close()\n",
    "file2.close()\n",
    "file3.close()\n",
    "\n",
    "from itertools import combinations \n",
    "#msa\n",
    "sequences = [seq1, seq2, seq3]\n",
    "\n",
    "#calcualtes the combinations of 2\n",
    "seqs = combinations(sequences, 2) \n",
    "\n",
    "#variables to stores the outputs from the pairwise allignment\n",
    "seq_info = []\n",
    "identities = []\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "#Dynamic Sequencing\n",
    "def dynamic_seq(seq1, seq2):\n",
    "    #creates the matrixes\n",
    "    matrix = np.zeros((len(seq2)+1, len(seq1)+1))\n",
    "    mutation = []\n",
    "    for row in range(0, len(seq2)+1):\n",
    "        values = []\n",
    "        for col in range (0, len(seq1)+1):\n",
    "            values.append([])\n",
    "        mutation.append(values)\n",
    "\n",
    "\n",
    "\n",
    "    #scores for Needleman gap, mismatch, match\n",
    "    gap = -2\n",
    "    match = 1\n",
    "    mismatch = -1\n",
    "\n",
    "\n",
    "    #first fill in the first row and column\n",
    "    for row in range(1, len(seq2)+1):\n",
    "        matrix[row][0] = matrix [row-1][0]+gap\n",
    "        mutation [row][0].append(\"del\")\n",
    "    for col in range(1, len(seq1)+1):\n",
    "        matrix[0][col] = matrix [0][col-1]+gap\n",
    "        mutation [0][col].append(\"ins\")\n",
    "\n",
    "    #diagonal in the rest of the entries\n",
    "    for row in range (1, len(seq2)+1):\n",
    "        for col in range (1, len(seq1)+1):\n",
    "            value1 = 0\n",
    "            symbol = \"\"\n",
    "            #calculates values for match/mismatch\n",
    "            if seq1[col-1] == seq2[row-1]:\n",
    "                value1 = matrix[row-1][col-1] + match\n",
    "                symbol = \"match\"\n",
    "            else:\n",
    "                value1 = matrix[row-1][col-1] + mismatch\n",
    "                symbol = \"mismatch\"\n",
    "            #calculates values for gaps\n",
    "            value2 = matrix [row-1][col]+gap\n",
    "            value3 = matrix [row][col-1]+gap\n",
    "            matrix[row][col] = max(value1, value2, value3)\n",
    "\n",
    "            #assigns the symbol in the direction for where the value came from\n",
    "            if matrix[row][col] == value1:\n",
    "                mutation[row][col].append(symbol)\n",
    "            elif matrix[row][col] == value2:\n",
    "                mutation[row][col].append(\"del\")\n",
    "            elif matrix[row][col] == value3:\n",
    "                mutation[row][col].append(\"ins\")\n",
    "    return mutation\n",
    "\n",
    "#calculates the pairwise allignment \n",
    "def pairwise_allignment (seq1, seq2, mutation):\n",
    "    align1 = \"\"\n",
    "    align2 = \"\"\n",
    "    final = \"\"\n",
    "    i = len(seq2)\n",
    "    j = len(seq1)\n",
    "    \n",
    "    #detemines the mutation and aligns the two sequences\n",
    "    while i > 0 and j > 0:\n",
    "        symbols = mutation[i][j]\n",
    "        for symbol in symbols:\n",
    "            if symbol == \"mismatch\":\n",
    "                align1 += seq2 [i-1]\n",
    "                align2 += seq1 [j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "                final += \"*\"\n",
    "            elif symbol == \"match\":\n",
    "                align1 += seq2 [i-1]\n",
    "                align2 += seq1 [j-1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "                final += \" \"\n",
    "            elif symbol == \"del\":\n",
    "                align1 += seq2[i-1]\n",
    "                align2 += \"-\"\n",
    "                i -= 1\n",
    "                final += \"*\"\n",
    "            elif symbol == \"ins\":\n",
    "                align1 += \"-\"\n",
    "                align2 += seq1[j-1]\n",
    "                j -= 1\n",
    "                final += \"*\"\n",
    "   \n",
    "    #completes allignment in the ends\n",
    "    while j > 0:\n",
    "        align1 += \"-\"\n",
    "        align2 += seq2[j-1]\n",
    "        j -= 1\n",
    "        final += \"*\"\n",
    "    while i > 0:\n",
    "        align1 += seq1[i-1]\n",
    "        align2 += \"-\"\n",
    "        i -= 1\n",
    "        final += \"*\"\n",
    "\n",
    "    #reorients the alignments\n",
    "    align1 = align1 [::-1]\n",
    "    align2 = align2 [::-1]\n",
    "    final = final [::-1]\n",
    "    \n",
    "    #calcualtes the different percent_identities\n",
    "    percent_identity = round(final.count (\" \")/len(final)*100,2)\n",
    "    return {\"align1\": align1,\"align2\": align2, \"final\": final, \"percent_identity\":percent_identity}\n",
    "\n",
    "#once you get the combination, does the allignments and assigns the percent identities to identities\n",
    "for seq in seqs: \n",
    "    mut = dynamic_seq(seq[0], seq[1])\n",
    "    info = pairwise_allignment (seq[0], seq[1], mut)\n",
    "    seq_info.append(info)\n",
    "    identities.append(info.get(\"percent_identity\"))\n",
    "\n",
    "from statistics import median\n",
    "\n",
    "#does the last two alignment\n",
    "last_alignment = seq_info[identities.index(median(identities))]\n",
    "\n",
    "#hard-coded ATT which is the first three nucleotides in Wuhan strain\n",
    "##reason for having to start with the 1st nucleotide in this code; \n",
    "##why we did not do 3 different msa's each of 10,000 base pairs\n",
    "#gets the alginment for the Tor2 strain and the Wuhan strain\n",
    "if seq_info[identities.index(median(identities))].get(\"align1\")[:3] == \"ATT\":\n",
    "    tor2 = seq_info[identities.index(median(identities))].get(\"align2\")\n",
    "    wuhan = seq_info[identities.index(median(identities))].get(\"align1\")\n",
    "else: \n",
    "    tor2 = seq_info[identities.index(median(identities))].get(\"align1\")\n",
    "    wuhan = seq_info[identities.index(median(identities))].get(\"align2\")\n",
    "\n",
    "#determines which of the sequence is the Urbani sequence after doing the alignment\n",
    "urbani_1 = seq_info[identities.index(max(identities))].get(\"align1\")\n",
    "urbani_2 = seq_info[identities.index(max(identities))].get(\"align2\")\n",
    "\n",
    "#does the two alignments with both the potential urbani strains\n",
    "first_alignment_one = dynamic_seq(tor2, urbani_1)\n",
    "first_align_info_one = pairwise_allignment(tor2, urbani_1, first_alignment_one)\n",
    "first_alignment_two = dynamic_seq(tor2, urbani_2)\n",
    "first_align_info_two = pairwise_allignment(tor2, urbani_2, first_alignment_two)\n",
    "\n",
    "#determines highest percent identity of the two alignments\n",
    "if first_align_info_one.get(\"percent_identity\") > first_align_info_two.get(\"percent_identity\"):\n",
    "    first_align_info = first_align_info_two\n",
    "else: \n",
    "    first_align_info = first_align_info_one \n",
    "\n",
    "#the urbani strain is determined\n",
    "if first_align_info.get(\"align1\") == tor2:\n",
    "    urbani = first_align_info.get(\"align2\")\n",
    "else: \n",
    "    urbani = first_align_info.get(\"align1\")\n",
    "\n",
    "#calculates the final alignment\n",
    "i = 0\n",
    "final = \"\"\n",
    "final1 = first_align_info.get(\"final\")\n",
    "final2 = last_alignment.get(\"final\")\n",
    "while i < len(final1):\n",
    "    if final1[i] == \"*\" or final2[i] == \"*\":\n",
    "        final = final + \"*\"\n",
    "    else:\n",
    "        final = final + \" \"\n",
    "    i += 1\n",
    "\n",
    "#calculates percent identity\n",
    "percent_identity = round(final.count(\" \")/len(final)*100,2)\n",
    "\n",
    "#creates an output file that aligns the three sequences together\n",
    "output = open(\"output.txt\", \"w\")\n",
    "i = 0\n",
    "\n",
    "#determines reminder for the last line\n",
    "reminder = len(final) % 50\n",
    "\n",
    "#prints the allignments in sets of 50\n",
    "while i < len(final)-reminder:\n",
    "    output.write(\"Urbani: \" + urbani[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Tor2:   \" + tor2[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Wuhan:  \" + wuhan[i:i+50]+\" \"+ str(i+50)+ \"\\n\")\n",
    "    output.write(\"Final:  \" + final[i:i+50]+\" \"+ str(i+50)+ \"\\n \\n\")\n",
    "    i = i + 50\n",
    "\n",
    "#prints the last lines of the allignment\n",
    "output.write(\"Urbani: \" + urbani[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Tor2:   \" + tor2[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Wuhan:  \" + wuhan[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n\")\n",
    "output.write(\"Final:  \" + final[i:i+reminder]+\" \"+ str(i+reminder)+ \"\\n \\n\")\n",
    "\n",
    "#prints the percent identity\n",
    "output.write(\"Percent Identity: \" + str(percent_identity) + \"%\")\n",
    "output.close()"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
